Implement Generic OIDC Provider

whitej6 opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Updates social-auth-core to 4.3.0 and implemented a generic OIDC SSO provider. #1967 
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 - UI
 - SSO
##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
awx: 9.0.2.dev8705+gb26a515497
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
Tested with a generic development provider. Only implemented with minimal set of variables required by social-auth-core.
<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

jmcgill298 commented:
This also fixes an issue with using a load-balancer with backend servers using different ports

https://github.com/python-social-auth/social-core/issues/609

whitej6 commented:
@shanemcd I resolved linting and licensing errors in tests, schema I test looks to be pulling from S3 so I am guessing that is expected to fail since I am adding new endpoints and appears to be a stead state schema export. 

The only test I was unable to resolve was screen tests. My background is python and some hints if possible would be greatly appreciated as I am happy to have all pipelines passing before thorough review. I copied the same patterns from AzureAD but still get failures an am unsure why.

whitej6 commented:
@shanemcd was able to fix my issues with screen tests. Now all tests that should pass do pass and is ready for review. Only test expected to fail is API schema due to registering a new endpoint to settings.

john-westcott-iv commented on awx/ui/src/screens/Setting/OIDC/OIDCEdit/OIDCEdit.js:
```diff
@@ -0,0 +1,142 @@
+                  name="SOCIAL_AUTH_OIDC_SECRET"
+                  config={OIDC.SOCIAL_AUTH_OIDC_SECRET}
+                />
+                <ObjectField
```
Why an object field here instead of just an InputField? And, actually, I think we have a URL validator field. @keithjgrant may be able to point you towards that.

whitej6 replied:
Likely a copy pasta oversight, will move to correct field type.

whitej6 replied:
Reading through SharedField I believe I just needed to change to InputField and `type="url"`

john-westcott-iv replied:
This is all set.

john-westcott-iv commented on awx/sso/conf.py:
```diff
@@ -1215,6 +1215,44 @@ def _register_ldap(append=None):
+    allow_blank=True,
+    default='',
+    label=_('OIDC Provider URL'),
+    help_text=_('The URL for your OIDC provider, e.g.: http(s)://hostname/.'),
```
Can we change this help_test to be something like `The URL for your OIDC provider including the path up to /.well-known/openid-configuration`

whitej6 replied:
```suggestion
    help_text=_('The URL for your OIDC provider including the path up to /.well-known/openid-configuration'),
```

john-westcott-iv commented on awx/sso/conf.py:
```diff
@@ -1215,6 +1215,44 @@ def _register_ldap(append=None):
+    category=_('Generic OIDC'),
+    category_slug='oidc',
+)
+
```
There is one other setting we may want to think about adding here:
```
register(
    'SOCIAL_AUTH_OIDC_VERIFY_SSL',
    field_class=fields.BooleanField,
    default=True,
    label=_('Verify OIDC Provider Certificate'),
    help_text=_('Verify the OIDV provider ssl certificate.'),
    category=_('Generic OIDC'),
    category_slug='oidc',
)
```
I'm a little torn because we don't do this for other providers but we also occasionally get questions about how to connect to a provider with an untrusted cert. This field may be a quick and easy way for an admin to do some testing. @AlanCoding any thoughts on adding this field?

whitej6 replied:
To play devil's advocate, OIDC _should_ always be SSL and you should _never_ do an implicit trusting of any cert but instead you should install the cert so you can do an explicit trusting of said cert.

Happy to add it or happy to leave as is, will wait on AlanCoding's response.

john-westcott-iv replied:
I 100% agree that OIDC should always be SSL and trusted in production. But people also use AWX for testing/development and personal use. We also allow credentials and other items to ignore SSL verification. And the stack trace when SSL is not trusted is pretty massive. I talked with an architect this morning and they are good with us adding this. If we wanted to we could make this just settable via the API and not have a UI field on the form.

john-westcott-iv replied:
Let's ignore this for now in this PR and I can add this as an option later as needed.

whitej6 replied:
@john-westcott-iv I just added it..... 😆 Want me to revert?

john-westcott-iv replied:
If you got it in there, its totally fine; thanks!
We'll get this over to our QE team and hopefully we can merge soon.

john-westcott-iv reviewed (commented):
Thanks again for the submission! Could you rebase this change on the latest devel branch along with some minor teaks mentioned below?

whitej6 commented:
@john-westcott-iv I fixed the two issues, pending to head if ssl verify is needed. Also rebased, got a little wonky in the process but I think everything is good. Let me know if there is anything else

john-westcott-iv commented on awx/ui/src/screens/Setting/OIDC/OIDCEdit/OIDCEdit.js:
```diff
@@ -0,0 +1,146 @@
+<<<<<<< HEAD
+=======
+  ObjectField,
+>>>>>>> 3505092cf8d127bc75dfe83d39f196a889478890
```
Looks like there was an unresolved merge issue here. I think these lines can just be removed.

whitej6 replied:
Sorry about that. Fixed.

jay-steurer commented:
Hi @whitej6. Could you rebase your PR? Therefore I can properly check it. Thank you!

whitej6 committed:
rebasing

whitej6 commented:
@jay-steurer rebased again 😃

jay-steurer commented:
Verified by QE.
